### DISCUSSION TITLE
OIDC Token Endpoint appears not to be sent Client ID / Secret

### DISCUSSION THREAD

**Discussion by gerryfletch** [pocketbase/pocketbase#4311] [Q&A]
Hi,

Two weeks ago I downloaded the pocketbase binary and configured the OIDC provider to interface with our authorization server. I have ran the same binary, with the same FE code using the all-in-one OAuth function provided by the Pocketbase JS SDK.

I'm rerunning it now, and I'm getting back a 401 because the authorization header is missing on the `/token` request: 

```
ERROR POST /api/collections/users/auth-with-oauth2
└─ Something went wrong while processing your request.
   └─ oauth2: "Invalid value for: header Authorization (missing)"
   ```
   
 I can't for the life of me figure out why. The auth server has always required the client ID and secret present in the authorization server using Basic auth so I can't figure out what's changed between then and now.

I tried configuring a second OIDC provider using a different client id / password pair. The client ID is passed along to the `/authorize` route correctly so PB has definitely picked up the settings.

Any ideas?!

**Comment by ganigeorgiev** [comment]
All OIDC providers are using the same implementation so I have no idea why it will work with one and not the other.

I'm not sure how to help you debug it as there are too many factors in play.

It could be something related to your authorization server, it could be a reverse proxy configuration, it could be cache related, etc.

I'm not ruling out a PocketBase issue but I have no idea how to reproduce it.
If you think that it is PocketBase related, please try to provide more details about your setup (ex. what is the authorization server you are using), how you are sending the `auth-with-oauth2` request and what version of the PocketBase executable and the SDK you are using.

**Reply by gerryfletch** [reply]
Hi @ganigeorgiev thank you for the prompt reply. I agree it is exceptionally unlikely for it to be a problem with pocketbase, I'm just not sure what else to test!

Here is my usage of the JS SDK:

```typescript
loginBtn.addEventListener("click", () => {
  const pb = new PocketBase('http://localhost:8000');
  pb.authStore.clear();
  const w: Window = window.open()!!;

  pb.collection('users').authWithOAuth2({ provider: 'oidc2', scopes: ["openid"], urlCallback: async (url) => {
    w.location.href = url;
    w.focus();
  } }).then(result => {
    console.log("Logged in");
    console.log(result);
    text.textContent = JSON.stringify(result)
  });
});
```

I can confirm that the same requests but with the client id and secret in the Authorization Basic header succeed, by sending the requests myself. And as mentioned, the `/authorize` call includes the clientId in the URI params.

I'm trying to see if I can get some more verbose logging on that error from the auth server, but of course it implies the header is not present at all (as opposed to an invalid header).

I tried looking at the PB code myself but since I'm not familiar with Go I can't see for sure whether the client id and secret configured in the admin console are always passed on for token code exchange? Are you able to confirm that's the case for me?

**Reply by ganigeorgiev** [reply]
> I can't see for sure whether the client id and secret configured in the admin console are always passed on for token code exchange? Are you able to confirm that's the case for me?

Yes, testing locally (_with PocketBase v0.21.2 and JS SDK v0.21.1_) the provider settings are properly loaded for me. You can find the code in: https://github.com/pocketbase/pocketbase/blob/8f8a7c3268ba8040c827b5c8f9acaf833cc910ca/forms/record_oauth2_login.go#L137-L161

> I can confirm that the same requests but with the client id and secret in the Authorization Basic header succeed, by sending the requests myself

If you are behind a reverse proxy double check that all header are passed without modifications.
Also note that by default the auth header style is left to "auto detect" which means that it will try both as POST `application/x-www-form-urlencoded` and HTTP Basic Authorization header (see the standard Go [`oauth2.Endpoint` godoc](https://pkg.go.dev/golang.org/x/oauth2#Endpoint)).


Try to isolate any other external factors and run only the executable. If you are using the prebuilt executable you can try placing the below static html in your `pb_public/index.html` and navigate to `http://127.0.0.1:8090` in your browser:

```html
<!DOCTYPE html>
<html>
<head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <title>OAuth2 test</title>
</head>
<body>
    <button id="btn">OAuth2 login</button>

    <script type="module">
        import PocketBase from "https://cdn.jsdelivr.net/gh/pocketbase/js-sdk@0.21.1/dist/pocketbase.es.mjs"

        const pb = new PocketBase("http://127.0.0.1:8090");

        document.getElementById("btn").addEventListener("click", () => {
            pb.collection("users").authWithOAuth2({
                provider: "oidc",
                // scopes: ["openid"], // enable if you need to
            }).then((data) => {
                console.log("Success:", data);
            }).catch((err) => {
                console.log("Failure:", err);
            });
        })
    </script>
</body>
</html>
```

If you need further help, please provide details what is the version of the PocketBase and the JS SDK you are using.

**Reply by gerryfletch** [reply]
Very interesting. We require the auth header as opposed to encoded form, I wonder if the error we're responding with isn't being treated properly by the OAuth2 lib and it's not swapping the auth style. This gives me something to work with - thank you.

FWIW, identical error with the `pb_public` version too 😢

**Reply by ganigeorgiev** [reply]
It is possible that the auth style detection is failing, I'm not sure. Looking at the Go source code it will try both but if your authorization server for some reason is not responding with an error I guess it could fail.

You can try to test this by enforcing it manually:

0. [Install Go 1.21+](https://go.dev/doc/install)
1. Clone/download the PocketBase repo
3. Edit `tools/auth/base_provider.go` and add in https://github.com/pocketbase/pocketbase/blob/8f8a7c3268ba8040c827b5c8f9acaf833cc910ca/tools/auth/base_provider.go#L188
  Try consequently with `AuthStyle: oauth2.AuthStyleInParams` and `AuthStyle: oauth2.AuthStyleInHeader`.

4. To test run the change, navigate to `example/base` and execute `go run main.go serve`.

**Reply by gerryfletch** [reply]
Thanks @ganigeorgiev for all the help, it's super appreciated, especially since I'm sure this isn't related to pocketbase anyway. I'll close off this discussion now, thank you!
